### PR TITLE
soundwire: intel: move sdw_prepare_stream to hw_params

### PR DESCRIPTION
**Pull request by bardliao**
sdw_deprepare_stream() is called by intel_hw_free(), but
sdw_prepare_stream() is called by intel_prepare(). That was not
symmetrical. Move sdw_prepare_stream() to intel_hw_params() can
ensure sdw_deprepare_stream() and sdw_prepare_stream() will be
called symmetrically.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

### CODE REVIEW

File: drivers/soundwire/intel.c [hunk 1]
```diff
@@ -835,7 +837,7 @@ static int intel_prepare(struct snd_pcm_substream *substream,
 		return -EIO;
 	}
 
-	return sdw_prepare_stream(dma->stream);
+	return 0;
```
Comment (RanderWang) [inline drivers/soundwire/intel.c]: you could delete intel_prepare if you do this

**Comment by RanderWang** [comment]
For pause-resume case, hw_free  is not invoked, so sdw_deprepare_stream() is not called, and
sdw_prepare_stream() will return error for this case.

hi @bardliao ,  how about to mov sdw_deprepare_stream() to trigger stop ? for this case, is there a trigger stop event ?

File: drivers/soundwire/intel.c (hunk 1 above)
Comment (bardliao) [reply drivers/soundwire/intel.c]: @RanderWang Yes, thanks.

**Comment by bardliao** [comment]
> For pause-resume case, hw_free is not invoked, so sdw_deprepare_stream() is not called, and
> sdw_prepare_stream() will return error for this case.

@RanderWang Sorry, I didn't get you here. For pause-resume case, hw_parames will not be invoked, right?

> 
> hi @bardliao , how about to mov sdw_deprepare_stream() to trigger stop ? for this case, is there a trigger stop event ?

Yes, there is a trigger event. But I am not sure if it will be too late if we call sdw_prepare_stream() in the trigger event instead of before it.

**Comment by plbossart** [comment]
> > For pause-resume case, hw_free is not invoked, so sdw_deprepare_stream() is not called, and
> > sdw_prepare_stream() will return error for this case.
> 
> @RanderWang Sorry, I didn't get you here. For pause-resume case, hw_parames will not be invoked, right?
> 
> > hi @bardliao , how about to mov sdw_deprepare_stream() to trigger stop ? for this case, is there a trigger stop event ?
> 
> Yes, there is a trigger event. But I am not sure if it will be too late if we call sdw_prepare_stream() in the trigger event instead of before it.

Do we actually support the pause/resume? we don't with HDaudio links so why should be try to be smart here?
Even if we are, it's better to leave the bits allocated and the stream prepared, so that the resume operation is just a channel enable. Prepare might take time and it's not quite compatible with the notion that triggers are immediate.

**Comment by ranj063** [comment]
@bardliao does this affect suspend/resume? prepare will get called upon resume but hw_params will not be.

**Comment by bardliao** [comment]
> @bardliao does this affect suspend/resume? prepare will get called upon resume but hw_params will not be.

@ranj063 You are right. Move prepare to hw_params may not be a good idea. However, currently,
[bus->params.bandwidth += m_rt->stream->params.rate *	m_rt->ch_count * m_rt->stream->params.bps;](https://github.com/thesofproject/linux/blob/integration/soundwire-alpha-rc2/drivers/soundwire/stream.c#L1485) and [bus->params.bandwidth -= m_rt->stream->params.rate * m_rt->ch_count * m_rt->stream->params.bps;](https://github.com/thesofproject/linux/blob/integration/soundwire-alpha-rc2/drivers/soundwire/stream.c#L1726) are not symmetrical. Maybe we should separate this part from the prepare/deprepare function and do it somewhere else.

**Comment by bardliao** [comment]
relpaced by https://github.com/thesofproject/linux/pull/1166

**Pull request closed without merging**